agent/spec/lib/kontena/workers/service_pod_worker_spec.rb

```diff
@@ -4,18 +4,101 @@
-      'updated_at' => Time.now.to_s, 'deploy_rev' => Time.now.to_s
+      'id' => 'foo/2',
+      'instance_number' => 2,
+      'updated_at' => Time.now.to_s,
+      'deploy_rev' => Time.now.to_s,
     )
   end
   let(:subject) { described_class.new(node, service_pod) }
 
-  describe '#ensure_desired_state' do
-    before(:each) do
-      mock_rpc_client
-      allow(rpc_client).to receive(:request)
+  describe '#apply' do
+    let(:container_id) { '8919f1a9fd05a0730a4b36549771e25895d71b9dd54c426bd6f5be969d39773c' }
+    let(:overlay_ip) { '10.81.128.1' }
+    let(:container_started_at) { Time.now }
+    let(:container) {double(:container, id: container_id,
+      name: 'foo-2',
+      overlay_ip: overlay_ip,
+      started_at: container_started_at,
+    ) }
+    let(:restarted_container) {double(:container, id: container_id,
+      name: 'foo-2',
+      overlay_ip: overlay_ip,
+      started_at: container_started_at + 1.0,
+    ) }
+
+    it 'ensures the container, and syncs to master' do
+      expect(subject.wrapped_object).to receive(:ensure_desired_state).and_return(container)
+      expect(subject.wrapped_object).to receive(:sync_state_to_master).with(service_pod, container)
+
+      subject.apply
+    end
+
+    it 'ensures the container, and syncs errors to master' do
+      expect(subject.wrapped_object).to receive(:ensure_desired_state).and_raise(RuntimeError.new('test'))
+      expect(subject.wrapped_object).to receive(:sync_state_to_master).with(service_pod, nil, RuntimeError)
+
+      subject.apply
     end
 
+    context 'with wait_for_port' do
+      let(:service_pod) do
+        Kontena::Models::ServicePod.new(
+          'id' => 'foo/2',
+          'instance_number' => 2,
+          'updated_at' => Time.now.to_s,
+          'deploy_rev' => Time.now.to_s,
+          'desired_state' => 'running',
+          'wait_for_port' => 1337,
+        )
+      end
+
+      it 'syncs state after waiting for the port' do
+        expect(subject.wrapped_object).to receive(:ensure_desired_state).and_return(container)
+        expect(subject.wrapped_object).to receive(:port_open?).with(overlay_ip, 1337, timeout: 1.0).and_return(true)
+        expect(subject.wrapped_object).to receive(:sync_state_to_master).with(service_pod, container)
+
+        subject.apply
+      end
+
+      it 'aborts if the container crashes and restarts' do
+        expect(subject.wrapped_object).to receive(:ensure_desired_state).once.and_return(container)
+        expect(subject.wrapped_object).to receive(:port_open?).with(overlay_ip, 1337, timeout: 1.0).once do
+          expect(subject.wrapped_object).to receive(:ensure_desired_state).once.and_return(restarted_container)
+          expect(subject.wrapped_object).to receive(:wait_for_port)
+          # XXX: in this case both the failing initial wait_for_port, and the restart will report state...
+          expect(subject.wrapped_object).to receive(:sync_state_to_master).with(service_pod, restarted_container)
```

Comment:
If the restarted container completes `wait_for_port` immediately without sleeping, then it might report `state: 'running'` right away. The initial `apply` -> `wait_for_port` task will then race with that to report `error: "container restarted"`, because the `@service_pod.deploy_rev` is still the same.